### PR TITLE
Add services to store users wallet in local storage

### PR DESCRIPTION
**Pull request by Marik-D**
Contains #855

### CODE REVIEW

File: universal-login-wallet/test/integration/helpers/chaiConfig.ts [hunk 1]
```diff
@@ -0,0 +1,6 @@
+import chai from 'chai';
+import sinonChai from 'sinon-chai';
+import {solidity} from 'ethereum-waffle';
+
+chai.use(solidity);
+chai.use(sinonChai);
```
Comment (marekkirejczyk) [inline universal-login-wallet/test/integration/helpers/chaiConfig.ts]: That should probably be a separate PR.

Comment (marekkirejczyk) [reply universal-login-wallet/test/integration/helpers/chaiConfig.ts]: Is it used anywhere?

File: universal-login-wallet/src/ui/react/App.tsx [hunk 2]
```diff
@@ -19,7 +19,18 @@ import {ConnectionNotificationScreen} from './ConnectAccount/ConnectionNotificat
...
+  const {walletService, walletStorageService, sdk} = useServices();
+  const [appReady, setAppReady] = useState(false);
+
+  useLayoutEffect(() => {
+    walletStorageService.load();
+    setAppReady(true);
+  }, []);
+
+  if (!appReady) {
+    return null;
```
Comment (marekkirejczyk) [inline universal-login-wallet/src/ui/react/App.tsx]: Shall we do `return 'Loading...';` instead?

File: universal-login-wallet/src/core/services/StorageService.ts [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+}
+
+export class StorageEntry<T> {
+  constructor(
+    private key: string,
+    private sanitizer: Sanitizer<T>,
+    private storageService: StorageService,
+  ) {
+  }
+
```
Comment (marekkirejczyk) [inline universal-login-wallet/src/core/services/StorageService.ts]: What's with the newlines? We should always use one empty line between methods.

File: universal-login-wallet/src/core/services/StorageService.ts [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+
+  get(): T | null {
+    const json = this.storageService.get(this.key);
+    if (!json) {
+      return null;
+    }
+
```
Comment (marekkirejczyk) [inline universal-login-wallet/src/core/services/StorageService.ts]: Empty newline in a method? Meh.

File: universal-login-wallet/src/core/services/StorageService.ts [hunk 5]
```diff
@@ -0,0 +1,35 @@
...
+  get(key: string): string | null {
+    return localStorage.getItem(key);
+  }
+
+  set(key: string, value: string) {
+    localStorage.setItem(key, value);
+  }
+}
+
+export class StorageEntry<T> {
```
Comment (marekkirejczyk) [inline universal-login-wallet/src/core/services/StorageService.ts]: Abstraction level 10000 👍 😄

File: universal-login-wallet/src/core/services/WalletStorageService.ts [hunk 6]
```diff
@@ -0,0 +1,40 @@
...
+import {WalletService} from '@universal-login/sdk';
+import {ApplicationWallet, ensure} from '@universal-login/commons';
+import {asObject, asString, Sanitizer} from '@restless/sanitizers';
+
+export class WalletStorageService {
+  private storage: StorageEntry<ApplicationWallet>;
+
+  constructor(
+    private walletService: WalletService,
+    storageService: StorageService,
```
Comment (marekkirejczyk) [inline universal-login-wallet/src/core/services/WalletStorageService.ts]: We should make a maximum length of a line of 120 characters. Simple function headers take to many lines.

File: universal-login-wallet/src/core/services/WalletStorageService.ts [hunk 7]
```diff
@@ -0,0 +1,40 @@
+import {StorageEntry, StorageService} from './StorageService';
+import {WalletService} from '@universal-login/sdk';
+import {ApplicationWallet, ensure} from '@universal-login/commons';
+import {asObject, asString, Sanitizer} from '@restless/sanitizers';
+
+export class WalletStorageService {
+  private storage: StorageEntry<ApplicationWallet>;
+
+  constructor(
+    private walletService: WalletService,
```
Comment (marekkirejczyk) [inline universal-login-wallet/src/core/services/WalletStorageService.ts]: Shouldn't dependency be another way around? i.e. WalletService depend on StorageService?

File: universal-login-wallet/src/ui/react/ConnectAccount/ConnectWithPassphrase.tsx [hunk 8]
```diff
@@ -11,13 +11,14 @@ export const ConnectWithPassphrase = ({name}: ConnectWithPasssphraseProps) => {
...
   const [errorMessage, setErrorMessage] = useState('');
-  const {walletService} = useServices();
+  const {walletService, walletStorageService} = useServices();
   const {history} = useRouter();
 
   const onRecoveryClick = async () => {
     setIsLoading(true);
     try {
       await walletService.recover(name, code);
+      walletStorageService.save();
```
Comment (marekkirejczyk) [inline universal-login-wallet/src/ui/react/ConnectAccount/ConnectWithPassphrase.tsx]: Should be just one line: 
await walletService.recover(name, code);
#reversedependencies

File: universal-login-wallet/src/ui/react/CreateAccount/CreateAccount.tsx [hunk 9]
```diff
@@ -23,6 +23,7 @@ export const CreateAccount = ({location}: CreateAccountProps) => {
     modalService.showModal('waitingForDeploy');
     await deploy(name, defaultDeployOptions.gasPrice.toString());
     walletService.setDeployed(name);
+    walletStorageService.save();
```
Comment (marekkirejczyk) [inline universal-login-wallet/src/ui/react/CreateAccount/CreateAccount.tsx]: Should be just one line. #reversedependencies

File: universal-login-wallet/src/core/services/WalletStorageService.ts [hunk 10]
```diff
@@ -0,0 +1,40 @@
...
+    const wallet = this.walletService.applicationWallet as ApplicationWallet;
+    this.storage.set(wallet);
+  }
+
+  load() {
+    const wallet = this.storage.get();
+    if (wallet === null) {
+      return false;
+    }
+    this.walletService.connect(wallet);
```
Comment (marekkirejczyk) [inline universal-login-wallet/src/core/services/WalletStorageService.ts]: #reversedependencies

File: universal-login-wallet/src/core/services/WalletStorageService.ts [hunk 11]
```diff
@@ -0,0 +1,40 @@
...
+      'wallet',
+      asApplicationWallet,
+      storageService,
+    );
+  }
+
+  save() {
+    ensure(this.walletService.isAuthorized(), Error, 'User must be authorized');
+    const wallet = this.walletService.applicationWallet as ApplicationWallet;
+    this.storage.set(wallet);
```
Comment (marekkirejczyk) [inline universal-login-wallet/src/core/services/WalletStorageService.ts]: #reversedependencies

File: universal-login-wallet/test/integration/helpers/chaiConfig.ts (hunk 1 above)
Comment (Marik-D) [reply universal-login-wallet/test/integration/helpers/chaiConfig.ts]: Yep, it's used in `LoginServices` test

File: universal-login-wallet/src/ui/react/App.tsx (hunk 2 above)
Comment (Marik-D) [reply universal-login-wallet/src/ui/react/App.tsx]: This will result in a flash of text in this case. I've only added it here since the wallet service is queried directly, which means that it wouldn't update the component by itself. So we need a state just to trigger a re-render.

File: universal-login-wallet/src/core/services/StorageService.ts (hunk 5 above)
Comment (Marik-D) [reply universal-login-wallet/src/core/services/StorageService.ts]: `StorageEntry` is just a nice interface that bundles up together a storage key and a sanitizer for a specific value to prevent mixups and type-unsafe value casts.

File: universal-login-wallet/src/core/services/WalletStorageService.ts (hunk 6 above)
Comment (Marik-D) [reply universal-login-wallet/src/core/services/WalletStorageService.ts]: In my opinion, even with 2 parameters, this variant is easier to read, since every parameter occupies a single line, it's easy to distinguish names from types. And you can just skim through it to know what that constructor takes.

File: universal-login-wallet/src/ui/react/App.tsx (hunk 2 above)
Comment (Marik-D) [reply universal-login-wallet/src/ui/react/App.tsx]: We should think about a way to add an ability to watch for changes in the wallet service.

File: universal-login-wallet/src/core/services/WalletStorageService.ts (hunk 7 above)
Comment (Marik-D) [reply universal-login-wallet/src/core/services/WalletStorageService.ts]: Yeah, I can add a wallet storage as a parameter to wallet service that defaults to in-memory storage

File: universal-login-wallet/src/ui/react/App.tsx (hunk 2 above)
Comment (marekkirejczyk) [reply universal-login-wallet/src/ui/react/App.tsx]: It is on the roadmap.

File: universal-login-sdk/lib/core/services/WalletService.ts [hunk 12]
```diff
@@ -8,11 +8,21 @@ type WalletState = 'None' | 'Future' | 'Deployed';
 
 type WalletFromBackupCodes = (username: string, password: string) => Promise<Wallet>;
 
+export interface WalletStorage {
+  load(): ApplicationWallet | null;
+
+  save(wallet: ApplicationWallet | null): void;
+}
```
Comment (marekkirejczyk) [inline universal-login-sdk/lib/core/services/WalletService.ts]: extract file?
